samples/java/src/main/java/com/gradle/enterprise/export/ExportApiJavaExample.java

```diff
@@ -1,119 +1,211 @@
+            JsonNode json = parse(data);
+            JsonNode buildToolJson = json.get("toolType");
+            if (buildToolJson == null || buildToolJson.asText().equals("gradle") || buildToolJson.asText().equals("maven")) {
+                final String buildId = json.get("buildId").asText();
+                final String buildTool = buildToolJson != null ? buildToolJson.asText() : "gradle";
+
+                Request request = requestBuildScan(buildId, buildTool);
+
+                ExtractUsernameFromBuildScan listener = new ExtractUsernameFromBuildScan(buildId);
+                eventSourceFactory.newEventSource(request, listener);
+                candiateUsernames.add(listener.getUsername());
+            }
+        }
 
-    private static Observable<ServerSentEvent> buildEventStream(String buildId, Set<String> eventTypes) {
-        return resume("/build-export/v2/build/" + buildId + "/events?eventTypes=" + Joiner.on(",").join(eventTypes), null);
+        @Override
+        public void onClosed(@NotNull EventSource eventSource) {
+            usernames.complete(candiateUsernames.stream()
+                    .map(u -> {
+                        try {
+                            return u.get();
+                        } catch (InterruptedException | ExecutionException e) {
+                            e.printStackTrace();
+                            return "Unknown";
+                        }
+                    })
+                    .collect(Collectors.toList())
+            );
+        }
     }
 
-    private static Observable<ServerSentEvent> resume(String url, String lastEventId) {
-        AtomicReference<String> eventId = new AtomicReference<>();
+    private static class ExtractUsernameFromBuildScan extends PrintFailuresEventSourceListener {
+        private final String buildId;
+        private final CompletableFuture<String> username = new CompletableFuture<>();
+
+        private ExtractUsernameFromBuildScan(String buildId) {
+            this.buildId = buildId;
+        }
+
+        public CompletableFuture<String> getUsername() {
+            return username;
+        }
 
-        HttpClientRequest<ByteBuf, ByteBuf> request = HTTP_CLIENT
-            .createGet(url)
-            .addHeader("Authorization", "Basic " + EXPORT_API_CREDENTIALS);
+        @Override
+        public void onOpen(@NotNull EventSource eventSource, @NotNull Response response) {
+            System.out.println("Streaming events for : " + buildId);
+        }
 
-        if (lastEventId != null) {
-            request = request.addHeader("Last-Event-ID", lastEventId);
+        @Override
+        public void onEvent(@NotNull EventSource eventSource, @Nullable String id, @Nullable String type, @NotNull String data) {
+            if (type.equals("BuildEvent")) {
+                JsonNode eventJson = parse(data);
+                username.complete(eventJson.get("data").get("username").asText());
```

Comment:
FYI, for Bazel things are a bit more complex
The username can be retrieved from sth like
```
.onData(BzlWorkspaceStatusItem_1_0.class, event -> {
                        if (event.key.equals(BUILD_USER)) {
                        // the username is the value of this guy
                        }
                    })
```
Given we'll have to add Bazel here very soon, it would be nice if the code was already a bit ready.
I'm thinking that you could have a Map<BuildTool, Function<JsonNode,String>>, and each build tool would declare the function used to grab the username from the provided event json.
Of course, `ExtractUsernameFromBuildScan` will have to know the build tool type of the given buildId, to know which function to use.

Reply:
Good suggestion! I decided to create a BuildTool enum rather than introducing another map.